### ISSUE TITLE
batchnorm is applied after relu in the tensorflow implementation of ResNet18

### ISSUE THREAD

**Issue by lzx325** [srebuffi/iCaRL#5]
In the tensorflow version of ResNet18, the residual block is implemented as follows
```python
layer = conv(inp, 'resconv1'+nom, size=3, strides=first_stride, out_channels=out_num_filters, alpha=alpha, padding='SAME')
layer = batch_norm(layer, 'batch_norm_resconv1'+nom, phase=phase)
layer = conv(layer, 'resconv2'+nom, size=3, strides=[1, 1, 1, 1], out_channels=out_num_filters, apply_relu=False,alpha=alpha, padding='SAME')
layer = batch_norm(layer, 'batch_norm_resconv2'+nom, phase=phase)
```
(in utils_resnet.py:110-113)
where the first batch_norm layer is appended to the first conv layer(which is apply_relu=True by default).
Can you tell me why we apply batch_norm **after** relu here, which contradicts the orinary practice that applies batch_norm **before** relu?

**Comment by paganpasta** [comment]
> In the tensorflow version of ResNet18, the residual block is implemented as follows
> 
> ```python
> layer = conv(inp, 'resconv1'+nom, size=3, strides=first_stride, out_channels=out_num_filters, alpha=alpha, padding='SAME')
> layer = batch_norm(layer, 'batch_norm_resconv1'+nom, phase=phase)
> layer = conv(layer, 'resconv2'+nom, size=3, strides=[1, 1, 1, 1], out_channels=out_num_filters, apply_relu=False,alpha=alpha, padding='SAME')
> layer = batch_norm(layer, 'batch_norm_resconv2'+nom, phase=phase)
> ```
> 
> (in utils_resnet.py:110-113)
> where the first batch_norm layer is appended to the first conv layer(which is apply_relu=True by default).
> Can you tell me why we apply batch_norm **after** relu here, which contradicts the orinary practice that applies batch_norm **before** relu?

@lzx325  I don't know the author's reasoning behind the placement for batch-norm. Even though the original paper which proposed the batch-norm layer suggested to use it before the activation, in practice I have seen it being used after as well. The community seems to be divided on the "correct" placement.
